test/System/Config.cpp

```diff
@@ -0,0 +1,29 @@
+#include <SFML/Config.hpp>
+
+#include <doctest.h>
+
+TEST_CASE("SFML/Config.hpp")
+{
+    SUBCASE("Version macros")
+    {
+        CHECK(SFML_VERSION_MAJOR == 3);
+        CHECK(SFML_VERSION_MINOR == 0);
+        CHECK(SFML_VERSION_PATCH == 0);
+        CHECK_FALSE(SFML_VERSION_IS_RELEASE);
```

Comment:
Are tests of this form really useful? It's basically the exact same code as in `Config.hpp` and just means that when updating the version macros, it has to be done in two places.

In other words, what is this test trying to catch?

Reply:
It's making sure that more of the SFML API is correct and will remain correct. Do I expect this to ever be a source of bugs? No. But, it's easy to test and we ought to be testing as much of the API as possible. Just because certain aspects of the API are trivial doesn't mean we should intentionally never test them. 

These particular macros are actually never once used anywhere within SFML itself. If they got moved, removed, or their value changed, we have no way of knowing that through any other means than manual inspection which we know is prone to fail.

Reply:
One can argue that it's part of the public API and thus would make sense to test.  
On the other hand, the tests are rather questionable and need to be updated with every version change. 😄 

It doesn't hurt us that much, so I'm fine with keeping them.

Reply:
I'm open to the possibility that I may be wrong about this being worth testing but luckily it's trivial to remove at any point in the future if we discover more reasons why it's not a good idea.

Reply:
> Do I expect this to ever be a source of bugs? No. 

Then I would say there is no point in testing it. 

> But, it's easy to test and we ought to be testing as much of the API as possible.

It sounds philosophical, but why do we test? 😉

I personally would argue it's to prevent current and future bugs. So if there's very low probability of that happening, then tests lack a purpose in my opinion. There are of course other valid views, e.g. tests acting as a sort of API specification, or fulfilling a coverage metric, etc.

---

While I agree that tests are easy to write and thus tempting, you need to keep in mind each test comes with a tiny maintenance cost. Sure, adjusting two places with constants instead of one may only be a few seconds more, but the bigger issue is when a developer forgets it. Someone updates the `Config.hpp` version, pushes the code. Someone else notices that the CI fails and messages the first developer, who is already working on something else and has to context-switch again. Quickly half an hour wasted for no gain.

Of course I'm speculating, but things like these happen all the time. A lot of people forget that usual software engineering principles (e.g. don't repeat yourself, think about coupling) also apply to tests. It's just that in many cases, violations of them are tolerated because the benefits of having tests often outweigh the downsides.

The other question is, whether copy-pasting the implementation code in the test code is a good strategy. By design, such tests cannot find mistakes in the implementations because they're identical copies. The only thing they are good for is to detect _changes_ in implementation -- which then means the test will be updated to match the implementation 1:1 again, and we're back to square one.

Another issue I have with this test in particular is that it doesn't prevent the most likely scenario: a new SFML version is released and someone forgets to update the config header. Of course, the code alone cannot verify this. Maybe instead we can think of something on the "meta" level, e.g. a CI job that is run when a new Git tag is created and verifies that `Config.hpp` matches the tag's version?

Reply:
> The other question is, whether copy-pasting the implementation code in the test code is a good strategy.

I'm not sure where you got this idea of me pasting the implementation into the tests. That's never been the case.

> Another issue I have with this test in particular is that it doesn't prevent the most likely scenario: a new SFML version is released and someone forgets to update the config header.

Would you be more inclined to accept this if the tests used the CMake `PROJECT_VERSION` variables to ensure that the macros in Config.hpp always line up with what is specified in the call to `project`? I just pushed up this change. It means that if you change the version number whatsoever or change VERSION_IS_RELEASE without updating Config.hpp, the tests fail. I hadn't previously considering doing this but it does make these tests much much more valuable.

Reply:
> I'm not sure where you got this idea of me pasting the implementation into the tests. That's never been the case.

That was unfortunate wording of mine, sorry for that -- I didn't mean this as literal copy-pasting, let alone laziness (for this I've followed your contributions for too long 😉).

I meant that implementation code is semantically identical with test code (impl declares constant X with value Y, test checks if constant X has value Y).


> Would you be more inclined to accept this if the tests used the CMake `PROJECT_VERSION` variables to ensure that the macros in Config.hpp always line up with what is specified in the call to `project`?

Absolutely! This suddenly adds a lot of value because it makes sure two independent components stay in sync! 👍👍